Lib/shlex.py

```diff
@@ -312,26 +316,78 @@ def split(s, comments=False, posix=True):
     lex.whitespace_split = True
     if not comments:
         lex.commenters = ''
-    return list(lex)
+
+    if isinstance(s, bytes):
+        return [i.encode("ascii") for i in lex]
+    else:
+        return list(lex)
 
 
 def join(split_command):
     """Return a shell-escaped string from *split_command*."""
-    return ' '.join(quote(arg) for arg in split_command)
+    if len(split_command) == 0:
+        return ""
+
+    # Ensure all objects are the same type as the first one,
+    # otherwise convert and raise a warning
+    cleaned = []
+    warned = False
+
+    for command in split_command:
```

Comment:
I think it is too complicated. The str-version assumes that the objects are all strings, namely that `split_command` is an iterable of strings.